### PR TITLE
Issue #12164: Enable Optional Checker for detecting errors in using the Optional type

### PR DESCRIPTION
**Pull request by Vyom-Yadav**
#12164 

Enabled the [Optional Checker](https://checkerframework.org/manual/#optional-checker) for detecting errors in using the [Optional](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Optional.html) type.

### CODE REVIEW

**Comment by Vyom-Yadav** [comment]
```
Error:  /home/runner/work/checkstyle/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:[304,28] error: [method.invocation.invalid] call to <X>orElseThrow(java.util.function.Supplier<? extends X>) not allowed on the given receiver.
Error:    found   : @MaybePresent Optional</*INFERENCE FAILED for:*/ ? extends Object>
Error:    required: @Present Optional</*INFERENCE FAILED for:*/ ? extends Object>
Error:  /home/runner/work/checkstyle/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:[328,28] error: [method.invocation.invalid] call to <X>orElseThrow(java.util.function.Supplier<? extends X>) not allowed on the given receiver.
Error:    found   : @MaybePresent Optional</*INFERENCE FAILED for:*/ ? extends Object>
Error:    required: @Present Optional</*INFERENCE FAILED for:*/ ? extends Object>
```

Basically the optional can contain no value and the checker framework doesn't allow calling `orElseThrow(..)` on `@MaybePresent Optional`.
From Docs:
> the Optional Checker will issue a warning if you call [get](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Optional.html#get()) or [orElseThrow](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Optional.html#orElseThrow(java.util.function.Supplier)) on a @MaybePresent Optional receiver, because each of these methods throws an exception if the receiver is an absent Optional.

@nrmancuso @romani I don't think this is a bad practice (using `orElseThrow(..)`), I think this is a good practice, better than doing:
```java
if (status.isPresent()) {
    return status.get();
} else {
    throw new Exception(); 
}
```
Thoughts?

**Comment by nrmancuso** [comment]
> I think this is a good practice, better than doing…

Please explain why you think this is a good practice.

**Comment by Vyom-Yadav** [comment]
> > I think this is a good practice, better than doing…
> 
> Please explain why you think this is a good practice.

I meant to say it is better than:
```java
if (status.isPresent()) {
    return status.get();
} else {
    throw new Exception(); 
}
```
throwing exceptions in general is not the best idea, they should be rarely used as part of a program's normal flow.

**Comment by romani** [comment]
I am not sure that extra IF will make code better.

>or [orElseThrow](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Optional.html#orElseThrow(java.util.function.Supplier)) on a @MaybePresent Optional receiver, because each of these methods throws an exception if the receiver is an absent Optional.

whole idea of this method to throw exception if something is missed.
Can we contact checker maintainers or stackoverflow for details on why it is like this ?  or may be this is defect in Checker ?

**Comment by romani** [comment]
here is details - https://github.com/typetools/checker-framework/issues/2537

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: I am beginning to understand this Checker a bit more... our code does not make sense. If we are guaranteeing that `firstChildOfMatchingText` is present, then we should never need to call `orElseThrow`. Also, when we `map` to some other Optional, the return type of `map` is not `@Present`.

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: I agree with this Checker, but we have an issue; compilation is failing, so code must be fixed to activate Optional Checker.

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: After applying this patch:
```
➜  checkstyle git:(pull-12184) ✗ git diff | cat
diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java b/src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java
index fb54dba9c..cd8d99c40 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java
@@ -37,6 +37,8 @@ import java.util.stream.Collectors;
 import javax.xml.parsers.ParserConfigurationException;
 import javax.xml.transform.TransformerException;
 
+import org.checkerframework.checker.optional.qual.Present;
+
 import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
 import com.puppycrawl.tools.checkstyle.api.DetailAST;
 import com.puppycrawl.tools.checkstyle.api.DetailNode;
@@ -300,12 +302,12 @@ public class JavadocMetadataScraper extends AbstractJavadocCheck {
             final DetailNode propertyNameTag = propertyNameNode.get();
             final String propertyName = getTextFromTag(propertyNameTag);
 
-            final DetailNode propertyType = getFirstChildOfMatchingText(nodeLi, TYPE_TAG)
-                .orElseThrow(() -> {
-                    return new MetadataGenerationException(String.format(
-                        Locale.ROOT, PROP_TYPE_MISSING, propertyName)
-                    );
-                });
+            final DetailNode propertyType = getFirstChildOfMatchingText(nodeLi, TYPE_TAG);
+            if (propertyType == null) {
+               throw new MetadataGenerationException(String.format(
+                                        Locale.ROOT, PROP_TYPE_MISSING, propertyName));
+            }
+
             final String propertyDesc = DESC_CLEAN.matcher(
                     constructSubTreeText(nodeLi, propertyNameTag.getIndex() + 1,
                             propertyType.getIndex() - 1))
@@ -315,21 +317,23 @@ public class JavadocMetadataScraper extends AbstractJavadocCheck {
             modulePropertyDetails.setName(propertyName);
             modulePropertyDetails.setType(getTagTextFromProperty(nodeLi, propertyType));
 
-            final Optional<DetailNode> validationTypeNodeOpt = getFirstChildOfMatchingText(nodeLi,
-                VALIDATION_TYPE_TAG);
-            if (validationTypeNodeOpt.isPresent()) {
-                final DetailNode validationTypeNode = validationTypeNodeOpt.get();
+            final DetailNode validationTypeNode = getFirstChildOfMatchingText(nodeLi, VALIDATION_TYPE_TAG);
+            if (validationTypeNode != null) {
                 modulePropertyDetails.setValidationType(getTagTextFromProperty(nodeLi,
                     validationTypeNode));
             }
 
-            final String defaultValue = getFirstChildOfMatchingText(nodeLi, DEFAULT_VALUE_TAG)
-                .map(defaultValueNode -> getPropertyDefaultText(nodeLi, defaultValueNode))
-                .orElseThrow(() -> {
-                    return new MetadataGenerationException(String.format(
-                        Locale.ROOT, PROP_DEFAULT_VALUE_MISSING, propertyName)
-                    );
-                });
+            final DetailNode nodeChild = getFirstChildOfMatchingText(nodeLi, DEFAULT_VALUE_TAG);
+            final String defaultValue;
+            if (nodeChild != null) {
+                defaultValue = getPropertyDefaultText(nodeLi, nodeChild);
+            }
+            else {
+                throw new MetadataGenerationException(
+                        String.format(Locale.ROOT, PROP_DEFAULT_VALUE_MISSING, propertyName)
+                );
+            }
+
             if (!PROPERTIES_TO_NOT_WRITE.contains(defaultValue)) {
                 modulePropertyDetails.setDefaultValue(defaultValue);
             }
@@ -512,11 +516,12 @@ public class JavadocMetadataScraper extends AbstractJavadocCheck {
      * @param pattern pattern to match against
      * @return the first child node matching the condition
      */
-    private static Optional<DetailNode> getFirstChildOfMatchingText(DetailNode node,
+    private static DetailNode getFirstChildOfMatchingText(DetailNode node,
                                                                     Pattern pattern) {
         return Arrays.stream(node.getChildren())
                 .filter(child -> pattern.matcher(child.getText()).matches())
-                .findFirst();
+                .findFirst()
+                .orElse(null);
     }
 
     /**

```

This makes Checker happy other than warnings.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: this conflicting with ideas to not use `null` (in some languages it is not a keyword at all, to not let it be used at alll) at all and deal will Optional.
Such refactoring will lead us to remove Optional. Something is wrong here....

Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: If the goal is to activate the checker anyhow, then `@SuppressWarnings("optional:method.invocation.invalid")` will suppress the violations and build will pass.

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: >Such refactoring will lead us to remove Optional. Something is wrong here....

The issue is with the method, since we treat the return value as `@Present` in two cases, and `@MaybePresent` in one case.

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: We could also just use optional inline as opposed to the method, and treat each optional separately, with the two that should throw exception as `@Present`

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: TLDR;
- if we annotate with `@Present`, absence of value is exceptional
- if we annotate with `@MaybePresent` (or no annotation, as this is default), absence is not exceptional

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: > absence of value is exceptional

Is the "value" in this sentence the possible value inside `Optional` or are we talking about `Optional` itself?

Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: I don't think annotating with `@Present` and `@MaybePresent` will work here. `java.util.Optional#map(..)` returns an optional of type `@MaybePresent` and `findFirst()` also returns the same. So until we annotate those methods to return of type `@Present` we would get the error.

Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: > > absence of value is exceptional
> 
> Is the "value" in this sentence the possible value inside `Optional` or are we talking about `Optional` itself?

I think "value" inside the optional.

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: > I don't think annotating with `@Present` and `@MaybePresent` will work here. `java.util.Optional#map(..)` returns an optional of type `@MaybePresent` and `findFirst()` also returns the same. So until we annotate those methods to return of type `@Present` we would get the error.

The issue is that we use the return value if the method as two different “types”. We can remove method, inline code, and use the appropriate annotation.

Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: I am unable to get what you are trying to say, checker will flag even if we use the wrong annotation. eg-
```java
@Present final Optional<String> optional =
        anyThingThatWillReturnOptional(..)
        .map(..);
```
`map(..)` will return `@MaybePresent` and our variable is annotated with `@Present`, the checker will flag this.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: Can we suppress without annotation?

**Comment by romani** [comment]
If this check is controversial, can try less problematic rule ? ( Separate PR)

File: src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java
Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: We can skip this file by using `-AskipDefs`

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: Ok, let’s not block this PR. Let’s:
1. Open an issue for further discussion
2. Add suppression by file mentioned at https://github.com/checkstyle/checkstyle/pull/12184#discussion_r967801525
3. Add comment with “until…” above suppression

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/meta/JavadocMetadataScraper.java:309]: Sounds good

File: pom.xml
Comment (romani) [inline pom.xml:2412]: Please add here all Checkers in comment to let us see all, disabled until issue, so we on activation, move Checker out of comment block.

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: pom.xml
Comment (nrmancuso) [inline pom.xml:389]: @Vyom-Yadav why aren't we using newest version?

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Item:

File: pom.xml
Comment (romani) [reply pom.xml:389]: https://checkerframework.org/manual/#installation

3.25

Comment (romani) [reply pom.xml:389]: Hmmm, why we are not running our shell script for version check on each PR to prevent such problem?
Is there any objections?

Comment (romani) [reply pom.xml:389]: new CI idea is moved to https://github.com/checkstyle/checkstyle/issues/12193

Comment (Vyom-Yadav) [reply pom.xml:389]: We aren't using the latest version because of 
```cmnd
[WARNING] 
Dependency convergence error for org.checkerframework:checker-qual:jar:3.12.0:compile paths to dependency are:
+-org.hibernate.search:hibernate-search-build-config:jar:6.2.0-SNAPSHOT
  +-com.puppycrawl.tools:checkstyle:jar:10.3.4-SNAPSHOT:compile
    +-com.google.guava:guava:jar:31.1-jre:compile
      +-org.checkerframework:checker-qual:jar:3.12.0:compile
and
+-org.hibernate.search:hibernate-search-build-config:jar:6.2.0-SNAPSHOT
  +-com.puppycrawl.tools:checkstyle:jar:10.3.4-SNAPSHOT:compile
    +-org.checkerframework:checker-qual:jar:3.25.0:compile

[ERROR] Rule 0: org.apache.maven.plugins.enforcer.DependencyConvergence failed with message:
Failed while enforcing releasability. See above detailed error message.
```

Comment (romani) [reply pom.xml:389]: such problems should not block us to use latest version.
We can make version of checker-qual as variable and set it to 3.12.0 before execution of hibernate.
Or we can ask hibernate project to make variable on their side to let us not conflict (we will do this after we become confident that we are going to use checker for sure).

Comment (Vyom-Yadav) [reply pom.xml:389]: Problem will occur in our codebase too if we use 3.25.0 without exclusions, I am closing this in favor of #12194, we can continue the discussion over there.

**Comment by Vyom-Yadav** [comment]
Closed in favour of #12194

**Pull request closed without merging**
